R/geneset_activity.R

```diff
@@ -0,0 +1,106 @@
+         character of gene IDs and the rest of the columns storing numeric
+         expression values for each sample.")
+  }
+
+  if (!check_input(model)) {
+    stop("The model should be a data.frame with the first column as a character
+         of gene IDs and the rest of the columns storing numeric weight values
+         for each node.")
+  }
+
+  if(!all(input_data[, 1] == model[, 1])){
+    stop("The gene IDs stored in the first column of the input data and the
+         first column of the model should be the same.")
+  }
+
+  weight_matrix <- as.matrix(model[, -1])
+  rownames(weight_matrix) <- model[[1]]
+  express_matrix <- as.matrix(input_data[, -1])
+
+  # calculate activities for each signature-geneset combination
+  activities <- sapply(1:nrow(signature_geneset_df), function(x) {
+    node <- as.numeric(gsub("\\D", "", signature_geneset_df[x, 1]))
+    side <- ifelse(grepl("pos", signature_geneset_df[x, 1]), "pos", "neg")
+    gene_set <- gene_set_list[[as.character(signature_geneset_df[x, 2])]]
+
+    one_signature_activity(weight_matrix = weight_matrix,
+                           express_matrix = express_matrix,
+                           node = node, side = side, gene_set = gene_set,
+                           HW_cutoff = HW_cutoff)
+  })
+
+  # transpose to have signatures in rows
+  activities <- dplyr::as_data_frame(t(activities))
+
+  # omit positive and negative signs of activities
+  activities <- abs(activities)
+
+  colnames(activities) <- colnames(input_data)[-1]
+
+  # add the signature name column in the front
+  activities <- data.frame(
+    signature = paste0(signature_geneset_df[, 1],
+                       "|", signature_geneset_df[, 2]),
+    activities, stringsAsFactors = FALSE, check.names = FALSE)
+
+  return(activities)
+
+}
+
+
+#' Combining geneset outputs
+#'
+#' Combines the signature geneset association output and the geneset limma
+#' output.
+#'
+#' @param signature_geneset_association a data.frame storing signatures and
+#' their significantly enriched gene sets, returned by the function
+#' annotate_signatures_with_genesets().
+#' @param geneset_limma_result a data.frame that stores the result table
+#' returned by limma. It includes logFC, adj.P.Val, and other statistics for
+#' each feature. The features are in the format "signature|geneset".
+#' @return a data.frame that stores both signature geneset association and
+#' genesets' limma result.
+#' @export
+combine_geneset_outputs <- function(signature_geneset_association,
+                                    geneset_limma_result){
+
+  # retrieve gene set and signature names
+  geneset_limma_result$signature <- sapply(
+    rownames(geneset_limma_result), function(x) unlist(strsplit(x, "\\|"))[[1]])
+  geneset_limma_result$geneset <- sapply(
+    rownames(geneset_limma_result), function(x) unlist(strsplit(x, "\\|"))[[2]])
+
+  # combine limma result with geneset association result
+  combined_result <- dplyr::left_join(
```

Comment:
i get what this is doing, just stumbled a bit through it. I think it would be easier to understand if you separate the dplyr chaining by what you want to join.

```R
this <- geneset_limma_result %>% dplyr::select(signature, geneset, logFC, adj.P.Val)
that <- signature_geneset_association %>% dplyr::select(-pvalue)
# I don't think need to specify signature = signature either?
combined_result <- dplyr::left_join(this, that, by = c('signature', 'geneset'))
```